pydantic/fields.py

```diff
@@ -945,10 +977,20 @@ class ComputedFieldInfo:
     alias_priority: int | None
     title: str | None
     description: str | None
+    deprecated: Deprecated | str | None
+    # This is necessary to avoid emitting two runtime warnings:
+    from_deprecated_decorator: bool
```

Comment:
@dmontagu,

Are we ok with having this as an additional param to the `ComputedFieldInfo` constructor? I can't think of an easy way to avoid having it here, and `ComputedFieldInfo` is a pretty internal data structure, so I think it's ok...

Reply:
I don't love that this is technically public API, but on the other hand it seems to me obviously intended for internal usage.

I think it may be a good idea to give these fields defaults to prevent theoretically breaking changes (and since there are obvious defaults — `deprecated=None` and `from_deprecated_decorator=False`). Or to make this class private (e.g. by adding a `__all__` and not including it). Probably making it explicitly non-public should be done in a separate PR.

Reply:
It looks to me like we can drop the `from_deprecated_decorator` with the following patch:
```diff
diff --git a/pydantic/_internal/_model_construction.py b/pydantic/_internal/_model_construction.py
index fac863bb2..4cd60874e 100644
--- a/pydantic/_internal/_model_construction.py
+++ b/pydantic/_internal/_model_construction.py
@@ -19,7 +19,7 @@ from ..errors import PydanticUndefinedAnnotation, PydanticUserError
 from ..plugin._schema_validator import create_schema_validator
 from ..warnings import GenericBeforeBaseModelWarning, PydanticDeprecatedSince20
 from ._config import ConfigWrapper
-from ._decorators import DecoratorInfos, PydanticDescriptorProxy, get_attribute_from_bases
+from ._decorators import DecoratorInfos, PydanticDescriptorProxy, get_attribute_from_bases, unwrap_wrapped_function
 from ._fields import collect_model_fields, is_valid_field_name, is_valid_privateattr_name
 from ._generate_schema import GenerateSchema
 from ._generics import PydanticGenericMetadata, get_model_typevars_map
@@ -584,8 +584,8 @@ def set_deprecated_descriptors(cls: type[BaseModel]) -> None:
 
     for field, computed_field_info in cls.model_computed_fields.items():
         if (
-            not computed_field_info.from_deprecated_decorator  # Avoid having two warnings emitted
-            and (msg := computed_field_info.deprecation_message) is not None
+            (msg := computed_field_info.deprecation_message) is not None
+            and not hasattr(unwrap_wrapped_function(computed_field_info.wrapped_property), '__deprecated__')
         ):
             desc = _DeprecatedFieldDescriptor(msg, computed_field_info.wrapped_property)
             desc.__set_name__(cls, field)
diff --git a/pydantic/fields.py b/pydantic/fields.py
index 8c344269e..ecea47d1b 100644
--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -979,7 +979,6 @@ class ComputedFieldInfo:
     description: str | None
     deprecated: Deprecated | str | None
     # This is necessary to avoid emitting two runtime warnings:
-    from_deprecated_decorator: bool
     examples: list[Any] | None
     json_schema_extra: JsonDict | typing.Callable[[JsonDict], None] | None
     repr: bool
@@ -1191,10 +1190,8 @@ def computed_field(
         if description is None and unwrapped.__doc__:
             description = inspect.cleandoc(unwrapped.__doc__)
 
-        from_deprecated_decorator = False
         if deprecated is None and hasattr(unwrapped, '__deprecated__'):
             deprecated = unwrapped.__deprecated__
-            from_deprecated_decorator = True
 
         # if the function isn't already decorated with `@property` (or another descriptor), then we wrap it now
         f = _decorators.ensure_property(f)
@@ -1213,7 +1210,6 @@ def computed_field(
             title,
             description,
             deprecated,
-            from_deprecated_decorator,
             examples,
             json_schema_extra,
             repr_,
```
I would propose we do so, just to keep the class definition a little cleaner. Since we have tests for the deprecation warning behaviors it seems okay to me.

Reply:
Related to the publicness of this API, the safest thing to do might be to make `deprecated` the last argument, with default `None`. But adding the default could make things weird in the future, and since there is some precedent for adding attributes to this class not at the end (specifically, #8013), maybe it is unnecessary.

I still think we should try to somehow mark this class as a private API so we can add things in the future without it technically being a breaking change.

Reply:
`unwrap_wrapped_function` makes things easier, nice.

Even if this wasn't public API, I wasn't really satisfied with the `from_deprecated_decorator` solution, which feels a bit dirty/hacky.